### PR TITLE
[Feat][Elementwise] Add DropoutKernel + DropoutOp with Philox PRNG

### LINKED ISSUE #444: [FEAT][ELEMENTWISE] implement dropout
## 1. Feature Scope

Dropout op with random mask generation. Different kernel pattern from standard elementwise — requires PRNG state management. Separate file `dropout.py`.

- `dropout(x, p, training, seed)` → masked and scaled output
- Deterministic replay: same seed → same mask
- Compatible with PyTorch's dropout semantics: `output = x * mask / (1 - p)` where mask is Bernoulli(1-p)

### Files

| File | Action |
|------|--------|
| `tileops/kernels/dropout.py` | New — DropoutKernel |
| `tileops/ops/dropout.py` | New — DropoutOp |
| `tileops/kernels/__init__.py` | Re-export |
| `tileops/ops/__init__.py` | Re-export |
| `tests/ops/test_dropout.py` | New |

### PRNG Design

Use Philox counter-based RNG (same as PyTorch CUDA):
- State: `(seed, offset)` — deterministic given same pair
- Per-element: `philox4x32(seed, offset + flat_idx)` → 4 random uint32 → convert to uniform float → threshold against `p`
- Each thread generates its own random numbers from deterministic state

### Kernel Sketch

```python
def _make_dropout_kernel(N, dtype, p, seed, offset, threads=256, npt=8):
    block_size = threads * npt
    scale = 1.0 / (1.0 - p)

    @T.prim_func
    def main(x: T.Tensor((N,), dtype), y: T.Tensor((N,), dtype)):
        with T.Kernel(T.ceildiv(N, block_size), threads=threads) as bx:
            for i, j in T.Parallel(threads, npt):
                idx = (bx * threads + i) * npt + j
                # Generate random value from Philox PRNG
                rand_val = philox_uniform(seed, offset + idx)
                mask = rand_val > p
                y[idx] = T.if_then_else(mask, x[idx] * scale, 0.0)
    return main
```

**Open question**: TileLang Philox support — may need to implement Philox rounds manually or use TileLang's random API if available.

### Correctness (3 cases)

| Test | Method |
|------|--------|
| Statistical distribution | Generate mask 100x, verify drop rate within 3σ of target p |
| Deterministic replay | Same seed → bitwise identical output |
| Scale factor | Non-dropped elements scaled by `1/(1-p)` |

### Edge Cases

| Case | Expected |
|------|----------|
| p=0.0 | Output = input (no dropout) |
| p=1.0 | Output = all zeros |
| training=False | Output = input (passthrough) |

## 4. Risks

| Risk | Mitigation |
|------|------------|
| TileLang may lack Philox PRNG primitives | Implement Philox 4-round counter manually in TileLang; or use `T.call_extern` for curand |
| Philox manual implementation correctness | Validate against PyTorch's `torch.cuda.manual_seed` + `F.dropout` output |
| Performance: per-element PRNG may be slow | Benchmark against PyTorch native; Philox is designed for GPU parallelism |

## 5. Acceptance Criteria

- [ ] DropoutKernel + DropoutOp implemented
- [ ] Deterministic replay verified (same seed = same output)
- [ ] Statistical drop rate within 3σ for p ∈ {0.1, 0.3, 0.5}
- [ ] Scale factor correct: non-dropped elements ×(1/(1-p))
- [ ] Edge cases pass: p=0, p=1, training=False
- [ ] `__init__.py` re-exports updated

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #444

## Summary

Implement dropout operator for TileOPs with Philox PRNG-based random mask generation. Adds `DropoutKernel` (tile-level kernel with Philox counter-based RNG) and `DropoutOp` (PyTorch-compatible op wrapper).

- **Philox PRNG**: deterministic, GPU-friendly — same seed produces identical masks
- **PyTorch semantics**: `output = x * mask / (1 - p)` where mask is Bernoulli(1 - p)
- **Edge cases**: p=0 (identity), p=1 (all zeros), training=False (passthrough)
- **torch.compile**: registered via `@torch.library.custom_op` + `register_fake`

**Supported dtypes**: `float16`, `bfloat16`, `float32`.

## Test plan

- [x] AC-1: DropoutKernel + DropoutOp implemented
- [x] AC-2: Deterministic replay — same seed = same output
- [x] AC-3: Statistical drop rate within 3σ for p ∈ {0.1, 0.3, 0.5}
- [x] AC-4: Scale factor correct — non-dropped elements scaled by 1/(1-p)
- [x] AC-5: Edge cases — p=0, p=1, training=False
- [x] AC-6: `__init__.py` re-exports updated
- [x] AC-7: torch.compile custom_op registered

**28 tests passed** (5.4s)

```bash
PYTHONPATH="$PWD" python -m pytest tests/ops/test_dropout.py -v
```

## Structural Compliance

All checks passed.

## Benchmark

**Environment**: NVIDIA H200, CUDA 12.8, PyTorch 2.9.1+cu128, TileLang 0.1.8

Shapes: (1024 × hidden_dim), p=0.5. Baseline: `F.dropout`.

| Shape | dtype | TileOPs (ms) | BW (TB/s) | Baseline (ms) | Baseline BW (TB/s) | Speedup |
|-------|-------|-------------|-----------|--------------|-------------------|---------|
| 1024×4096 | fp16 | 0.007 | 2.48 | 0.013 | 1.32 | 1.9x |
| 1024×4096 | bf16 | 0.007 | 2.48 | 0.013 | 1.31 | 1.9x |
| 1024×4096 | fp32 | 0.011 | 3.04 | 0.016 | 2.07 | 1.5x |
| 1024×10240 | fp16 | 0.013 | 3.21 | 0.024 | 1.75 | 1.8x |
| 1024×10240 | bf16 | 0.013 | 3.19 | 0.024 | 1.71 | 1.8x |
| 1024×10240 | fp32 | 0.023 | 3.62 | 0.034 | 2.46 | 1.5x |
| 1024×20480 | fp16 | 0.023 | 3.64 | 0.044 | 1.91 | 1.9x |
| 1024×20480 | bf16 | 0.023 | 3.63 | 0.045 | 1.87 | 2.0x |
| 1024×20480 | fp32 | 0.043 | 3.91 | 0.064 | 2.63 | 1.5x |

**Takeaways:**
- TileOPs dropout **1.5–2.0x faster** than `F.dropout` across all shapes/dtypes
- Bandwidth scales well: 2.5 → 3.6 → 3.9 TB/s (fp16) from small to large
- bf16 ≈ fp16; fp32 ~1.5x speedup (smaller gap due to higher per-element cost)
- TFLOPs not meaningful for dropout (RNG-dominated), omitted from table

**Benchmark command:**
```bash
PYTHONPATH="$PWD" python -m pytest benchmarks/ops/bench_dropout.py -v
# 9 passed in 71s
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/ops/test_dropout.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    )
+    y = kernel(x)
+
+    # Every element must be either 0 or scale
+    is_zero = y == 0
+    is_scaled = torch.isclose(
+        y.float(),
+        torch.full_like(y, scale, dtype=torch.float32),
+        atol=1e-3, rtol=1e-3,
+    )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_dropout.py:252]: Addressed in ebb1b90.

Evidence:
- Changed `atol=1e-3, rtol=1e-3` to `atol=1e-6, rtol=0` at line 257 of `tests/ops/test_dropout.py`
- The p=0.5 scale factor (2.0) is exactly representable, so the tighter tolerance catches regressions without false positives
- All 28 tests pass with the tighter tolerance

Current HEAD: a451785

File: tileops/kernels/__init__.py [hunk 2]
```diff
@@ -59,6 +60,7 @@
 
 __all__ = [
     "BinaryKernel",
+    "DropoutKernel",
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/__init__.py]: Addressed in a451785.

Evidence:
- Moved `DropoutKernel` to its correct alphabetical position in `__all__` (after `BatchNormFwdTrainKernel`, before `EngramDecodeKernel`)
- Verified via `git diff`: only the position of the `"DropoutKernel"` line changed within the `__all__` list

Current HEAD: a451785

File: tileops/ops/dropout.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        # Build kernel only when dropout is actually active
+        self._skip = not training or p == 0.0
+        if not self._skip:
+            self.dispatch_kernel(kernel_map)
+            self.kernel = self.kernel_map[self._op_name](
+                N_total, dtype, p=p, seed=seed, tune=tune,
+            )
+        else:
+            # Still need kernel_map for the interface but no kernel is built
+            self.kernel = None
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/dropout.py]: Addressed in ed61a31.

Evidence:
- Added `self._all_zero = training and p == 1.0` flag in `__init__`
- Kernel build condition changed to `if not self._skip and not self._all_zero` -- no JIT compilation when p=1.0
- `forward()` uses `self._all_zero` for the early return path
- Diff: `tileops/ops/dropout.py` lines 69-80

Current HEAD: ed61a31

File: tileops/ops/dropout.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/dropout.py]: Addressed in ed61a31.

Evidence:
- `dispatch_kernel(kernel_map)` is now called unconditionally (line 73 in new code), before the conditional kernel build
- `self.kernel_map` is always populated regardless of `_skip` or `_all_zero` state
- This matches the Op base class contract and allows introspection/debugging on all code paths

Current HEAD: ed61a31

File: tileops/kernels/dropout.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+)
+
+
+def _make_dropout_kernel(N, dtype, p, seed, threads=256, num_per_thread=8):
+    """Build a dropout kernel with Philox PRNG.
+
+    Each thread initializes its own Philox state using (seed, thread_global_id)
+    and generates uniform random floats. Elements where rand < p are dropped;
+    surviving elements are scaled by 1/(1-p).
+
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/dropout.py]: Addressed in ed61a31.

Evidence:
- Module docstring in `tileops/kernels/dropout.py` now says "using TileLang's T.rng_init / T.rng_rand_float (backed by cuRAND Philox4_32_10 by default)"
- Function docstring for `_make_dropout_kernel` updated: "Each thread calls T.rng_init(seed) which initialises per-thread RNG state"
- Class docstring for `DropoutKernel` updated: "Uses T.rng_init / T.rng_rand_float (backed by cuRAND Philox4_32_10 by default)"
- Class docstring for `DropoutOp` updated similarly
- Documentation now accurately describes the TileLang RNG API rather than claiming direct cuRAND Philox usage

Current HEAD: ed61a31

File: tileops/ops/dropout.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        self.seed = seed
+        self.training = training
+
+        # Build kernel only when dropout is actually active
+        self._skip = not training or p == 0.0
+        if not self._skip:
+            self.dispatch_kernel(kernel_map)
+            self.kernel = self.kernel_map[self._op_name](
+                N_total, dtype, p=p, seed=seed, tune=tune,
+            )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/dropout.py]: Addressed in ed61a31.

Evidence:
- Added `self._all_zero = training and p == 1.0` (line 70)
- Kernel build is now guarded by `if not self._skip and not self._all_zero` (line 76)
- When p=1.0, no JIT kernel compilation occurs -- `self.kernel = None` and `forward()` returns `torch.zeros_like(x)` via the `self._all_zero` early return

Current HEAD: ed61a31
